text/0000-glimmer-components.md

```diff
@@ -0,0 +1,835 @@
+users, since Ember has not had any modifiers besides `{{action}}`, and the
+element modifier manager RFC was just merged recently. Key things to focus on
+here will be the fact the modifiers run whenever an node is inserted or removed,
+so they may run multiple times during a components life (if they exist in an
+`{{if}}` or `{{each}}` block, for instance), and that they can receive pure
+function helpers if they do not require any state beyond the state of the
+element itself, which can simplify component code.
+
+### Converting From Legacy Components
+
+For users who are familiar with the legacy component API, some changes will be
+fairly simple, and others will be more nuanced. The individual focus points
+should be those outlined in the motivation of this RFC, and the motivation for
+the changes should be included in a summary that helps users understand the
+intent of new APIs so they can use them effectively.
+
+**Outer HTML semantics** should be fairly simple to communicate. These should
+also be the easiest to codemod, since it is mostly a 1-to-1 conversion from the
+legacy component definition to the template.
+
+**Namespaced arguments** will be a sticking point for some. The main thing to
+focus on here is aliasing to provide defaults, and the benefits of immutability.
+Patterns like the `@arg` decorator in
+[sparkles-decorators](https://github.com/gossi/sparkles-decorators) make this
+fairly ergonomic, and for simpler components defaulting using a class field
+initializer can also work.
+
+**Lifecycle hooks** may be fairly confusing to existing users, who are used to
+having a variety of hooks to choose from when coordinating lifecycle events. The
+bullet points here are:
+
+* `didCreate` is `didInsertElement`, without elements
+* `didUpdate` is the same as it is in legacy components
+* `destroy` is the correct place for all teardown code
+* `didReceiveAttrs`, `willRender`, and `didRender` code should generally be
+  extracted into a separate method called by both `didCreate` and `didUpdate`.
+* Any code that may cause a rerender in `didUpdate` or `didCreate` should
+  ideally be pushed into getters or helpers, so that it naturally is pulled into
+  the render instead of imperatively updated.
+
+**Property tracking** will be the most difficult change to understand in some
+regards, and will rely on dialing in the teaching for the mechanics of tracked
+properties (discussed in the tracked properties RFC).
+
+## Drawbacks
+
+The major drawback to Glimmer components is that they add a separate API for
+components, meaning that for the forseeable future Ember users will likely
+need to learn how to use both interchangeably. This introduces a fair amount of
+mental overhead for users, but the benefits of Glimmer components and their
+simplicity should make this less problematic.
+
+## Alternatives
+
+### No owner in constructor
+
+Sparkles components do not provide access to the owner or injections in the
```

Comment:
I just haven't implemented yet because it wasn't part of the "reference implementation" in glimmer.js. Sparkles will absolutely be updated (soon) to pass the owner in to the constructor so that other types of injections (e.g. `@service`) function properly.